File: educational_dp/ruby/v.rb
# V
# 全方位木DP

N, M = gets.split.map(&:to_i)
G = Array.new(N) { [] }
(N - 1).times do
  x, y = gets.split.map(&:to_i).map(&:pred)
  G[x] << y
  G[y] << x
end

@size = []
# posが黒色でかつ、黒色の頂点が全て連結である組み合わせを求める
def dfs1(pos, pre = -1)
  return @size[pos] if @size[pos]
  @size[pos] = 1
  G[pos].each do |i|
    next if i == pre
    # 1加算するのはpos以外白色だった場合を考慮
    @size[pos] *= dfs1(i, pos) + 1
    @size[pos] %= M
  end
  @size[pos]
end

@ans = []
def dfs2(pos, pre = -1)
  @ans[pos] = 1
  # l_cproduct, r_cproduct は累積積
  l_cproduct = []
  r_cproduct = []
  G[pos].each_with_index do |u, i|
    @ans[pos] *= @size[u] + 1
    @ans[pos] %= M
    l_cproduct[i] = r_cproduct[i] = (@size[u] + 1) % M
  end

  n = G[pos].size
  1.upto(n - 1) do |i|
    l_cproduct[i] *= l_cproduct[i - 1]
    l_cproduct[i] %= M
  end
  (n - 2).downto(0) do |i|
    r_cproduct[i] *= r_cproduct[i + 1]
    r_cproduct[i] %= M
  end

  G[pos].each_with_index do |u, i|
    next if u == pre
    # 親がuの場合の@size[pos]を計算する
    @size[pos] = 1
    @size[pos] *= l_cproduct[i - 1] unless i.zero?
    @size[pos] *= r_cproduct[i + 1] if i < n - 1
    @size[pos] %= M
    dfs2(u, pos)
  end
end

dfs1(0)
dfs2(0)
puts @ans

File: educational_dp/kotlin/v.kt
// V

fun solve() {
  val (N, M) = readLine()!!.split(" ").let { Pair(it[0].toInt(), it[1].toLong()) }
  val G = MutableList(N) { mutableListOf<Int>() }
  repeat(N - 1) {
    val (x, y) = readLine()!!.split(" ").map(String::toInt).map(Int::dec)
    G[x].add(y)
    G[y].add(x)
  }

  val dp = LongArray(N) { 0L }

  fun dfs1(pos: Int, pre: Int = -1): Long {
    if (dp[pos] != 0L) return dp[pos]
    dp[pos] = 1L
    G[pos].forEach {
      if (it == pre) return@forEach
      dp[pos] = dp[pos] * (dfs1(it, pos) + 1) % M
    }
    return dp[pos]
  }

  val ans = MutableList(N) { 0L }

  fun dfs2(pos: Int, pre: Int = -1) {
    ans[pos] = 1L
    G[pos].forEach {
      ans[pos] = ans[pos] * (dp[it] + 1) % M
      ans[pos] %= M
    }
    val n = G[pos].size
    val cProductL = MutableList(n) { 0L }
    val cProductR = MutableList(n) { 0L }
    G[pos].forEachIndexed { i, u ->
      cProductL[i] = dp[u] + 1
      cProductR[i] = dp[u] + 1
    }

    (1 .. n - 1).forEach {
      cProductL[it] = cProductL[it] * cProductL[it - 1] % M
      cProductR[n - it - 1] = cProductR[n - it - 1] * cProductR[n - it] % M
    }

    G[pos].forEachIndexed { i, u ->
      if (u == pre) return@forEachIndexed
      dp[pos] = 1L
      if (i != 0) dp[pos] = dp[pos] * cProductL[i - 1] % M
      if (i + 1 < n) dp[pos] = dp[pos] * cProductR[i + 1] % M
      dfs2(u, pos)
    }
  }

  dfs1(0)
  dfs2(0)

  ans.forEach(::println)
}

fun main() {
  Thread(null, ::solve, "solve", 1.shl(26)).start()
}
